src/gui/fspathedit.h

```diff
@@ -0,0 +1,151 @@
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#ifndef QBT_FSPATHEDIT_H
+#define QBT_FSPATHEDIT_H
+
+#include <QScopedPointer>
+#include <QWidget>
+
+namespace Private
+{
+    class FileEditorWithCompletion;
+    class FileLineEdit;
+    class FileComboEdit;
+}
+
+/*!
+ * \brief
+ * Widget for editing strings which are paths in filesystem
+ */
+class FileSystemPathEdit: public QWidget
+{
+    Q_OBJECT
+    Q_ENUMS(Mode)
+    Q_PROPERTY(Mode mode READ mode WRITE setMode)
+    Q_PROPERTY(QString selectedPath READ selectedPath WRITE setSelectedPath NOTIFY selectedPathChanged)
+    Q_PROPERTY(QString fileNameFilter READ fileNameFilter WRITE setFileNameFilter)
+    Q_PROPERTY(QString dialogCaption READ dialogCaption WRITE setDialogCaption)
+
+public:
+    ~FileSystemPathEdit() override;
+
+    enum class Mode
+    {
+        FileOpen,                                 //!< Working mode for opening files, shows open dialog
+        FileSave,                                 //!< working mode for saving files, shows save dialog
+        DirectoryOpen,                            //!< working mode for selecting existing directories
+        DirectorySave                             //!< working mode for selecting directories for saving
+    };
+
+    Mode mode() const;
+    void setMode(Mode mode);
+
+    QString currentDirectory() const;
+    QString selectedPath() const;
+    void setSelectedPath(const QString &val);
+
+    QString fileNameFilter() const;
+    void setFileNameFilter(const QString &val);
+
+    /// The browse button caption is "..." if true, and "Browse" otherwise
+    bool briefBrowseButtonCaption() const;
+    void setBriefBrowseButtonCaption(bool brief);
+
+    QString dialogCaption() const;
+    void setDialogCaption(const QString &caption);
+
+    virtual void clear() = 0;
+
+protected:
+    explicit FileSystemPathEdit(Private::FileEditorWithCompletion *editor, QWidget *parent);
+
+    template <class Widget>
+    Widget *editWidget() const
+    {
+        return static_cast<Widget *>(editWidgetImpl());
+    }
+
+protected slots:
+    void onPathEdited();
+
+signals:
+    void selectedPathChanged(const QString &path);
+
+private:
+    virtual QString editWidgetText() const = 0;
+    virtual void setEditWidgetText(const QString &text) = 0;
+
+    QWidget *editWidgetImpl() const;
+    Q_DISABLE_COPY(FileSystemPathEdit)
+    class FileSystemPathEditPrivate;
+    Q_DECLARE_PRIVATE(FileSystemPathEdit)
+    Q_PRIVATE_SLOT(d_func(), void browseActionTriggered())
```

Comment:
Qt has a naming convention for private slots (don't remember exactly how it looks).